centrallix-lib/src/xhash.c

```diff
@@ -290,4 +290,74 @@ xhClear(pXHashTable this, int (*free_fn)(), void* free_arg)
     return 0;
     }
 
+/*** Executes an operation on each entry of the hash table entry. 
+ *** 
+ *** @param this The affected hash table.
+ *** @param callback_fn A callback function to be called on each hash table
+ *** 	entry. It takes 2 parameters: the current hash table entry and a void*
+ *** 	argument specified using each_arg. If any invocation of the callback
+ *** 	function returns a value other than 0, xhForEach() will immediately
+ *** 	fail, returning that value as the error code.
+ *** @param each_arg An additional argument which will be passed to each
+ *** 	invocation of the callback function.
+ *** @returns 0 if the function executes successfully.
+ ***          1 if the callback function is NULL.
+ ***          n (where n != 0) if the callback function returns n.
+ ***/
+int
+xhForEach(pXHashTable this, int (*callback_fn)(pXHashEntry, void*), void* each_arg)
+    {
+    if (callback_fn == NULL) return 1;
+    
+    for (int row = 0; row < this->nRows; row++) 
+	{
+	pXHashEntry entry = (pXHashEntry)(this->Rows.Items[row]);
+	while (entry != NULL)
+	    {
+	    pXHashEntry next = entry->Next;
+	    const int ret = callback_fn(entry, each_arg);
+	    if (ret != 0) return ret;
+	    entry = next;
+	    }
+	}
+    
+    return 0;
+    }
+
+static int
+xhiFreeEntry(pXHashEntry entry, void* arg)
```

Comment:
Documentation for this and for xhClearKeySafe?

Also, if this is an "internal" function, naming convention should be xh_internal_FreeEntry() or xh_i_FreeEntry().  That separates the private function indicator from the prefix.